File: go/500.600/557.ReverseWordsInAString3.go
package main

import (
	"fmt"
)

func reverseWords(s string) string {
	byts := []byte(s)
	n := len(byts)
	p := 0
	for x := 0; x <= n; x++ {
		if x == n || byts[x] == ' ' {
			q := x - 1
			for p < q {
				byts[p], byts[q] = byts[q], byts[p]
				p++
				q--
			}
			p = x + 1
		}
	}
	return string(byts)
}

func main() {
	fmt.Println(reverseWords("Let's take LeetCode contest"), "s'teL ekat edoCteeL tsetnoc")
	fmt.Println(reverseWords("Mr Ding"), "rM gniD")
}

File: 500.600/557.ReverseWordsInAString3.cpp
#include <iostream>
#include <string>
using namespace std;

class ReverseWordsInAString3
{
public:
    string reverseWords(string s)
    {
        int n = s.size();
        int p = 0;
        for (int x = 0; x <= n; ++x)
        {
            if (x == n || isspace(s[x]))
            {
                int q = x - 1;
                while (p < q)
                {
                    swap(s[p], s[q]);
                    ++p;
                    --q;
                }
                p = x + 1;
            }
        }
        return s;
    }
};

int main()
{
    ReverseWordsInAString3 cls;
    cout << cls.reverseWords("Let's take LeetCode contest").c_str() << " s'teL ekat edoCteeL tsetnoc" << endl;
    cout << cls.reverseWords("Mr Ding").c_str() << " rM gniD" << endl;
    return 0;
}
